### PR TITLE
console.lua: refine the hovered item calculation

### PR DESCRIPTION
**Pull request by guidocella**
This is more accurate, especially with an always visible OSC bottombar.

### CODE REVIEW

File: player/lua/console.lua [hunk 1]
```diff
@@ -821,29 +822,26 @@ end
...
     end
 
-    if #matches < max_lines then
-        clicked_line = clicked_line - (max_lines - #matches)
-        max_lines = #matches
-    end
+    local last_match_to_print  = math.min(first_match_to_print + max_lines - 1,
+                                          #matches)
+
+    local hovered_item = last_match_to_print - math.floor((log_bottom_pos - y) / opts.font_size)
```
Comment (kasper93) [inline player/lua/console.lua:841]: I would prefer to use 2 for `log_bottom_pos` as we actually skip 2 lines and 0.5 here to offset half the line for selection round. It is bit hidden currently. 

For the record, I did my try on this before, and here is what I have that works well, feel free to use anything if you like
``` lua
local function determine_hovered_item()
    local height = select(2, get_scaled_osd_dimensions())
    local bottom_offset = height - mp.get_property_native('mouse-pos').y / scale_factor()
    -- Skip the input line and the empty line.
    local skip_lines = 2
    local bottom_margin = bottom_left_margin + global_margins.b * height + skip_lines * opts.font_size

    local mouse_over_line = math.floor((bottom_offset - bottom_margin) / opts.font_size + 0.5)
    if mouse_over_line < 0 then
        return
    end

    local max_lines = calculate_max_log_lines()

    -- Skip the counter if it's printed.
    if #matches > max_lines then
        max_lines = max_lines - 1
    end

    local visible_items = math.min(max_lines, #matches)
    if visible_items <= mouse_over_line then
        return
    end

    -- plus one for 1-based indexing
    return first_match_to_print + visible_items - (mouse_over_line + 1);
end
```

Comment (guidocella) [reply player/lua/console.lua:841]: I don't get it, where does 2 come from? You are adding 2 and then compensating for the error by rounding, while I calculate accurately with math.floor.

File: player/lua/console.lua [hunk 2]
```diff
@@ -821,29 +822,26 @@ end
 local function determine_hovered_item()
     local height = select(2, get_scaled_osd_dimensions())
     local y = mp.get_property_native('mouse-pos').y / scale_factor()
```
[the remark this answers is not in the document]
Comment (guidocella) [reply player/lua/console.lua:824]: It is flipped later with `log_bottom_pos - y`.

[the remark this answers is not in the document]
Comment (kasper93) [reply player/lua/console.lua:824]: Ok, effectively the same, just was easier to think to flip first to have absolute value and then convert to relative to log.

File: player/lua/console.lua (hunk 1 above)
Comment (kasper93) [reply player/lua/console.lua:841]: I don't get it, where does 1.5 come from? From my tests it is little over 1.6. Indeed `\N` is not full 1 in size.

Comment (guidocella) [reply player/lua/console.lua:841]: Well that's where it comes from? I also checked with `compute_bounds` and `\\N` is around half a line. See

```lua
mp.add_timeout(.1, function ()
    local result
    local ov = mp.create_osd_overlay('ass-events')
    ov.compute_bounds = true

    ov.data = 'foo\\Nbar'
    result = ov:update()
    print(result.y1 - result.y0)

    ov.data = 'foo\\N\\Nbar'
    result = ov:update()
    print(result.y1 - result.y0)
end)
```

Comment (kasper93) [reply player/lua/console.lua]: Why is this so ass? Not sure this test tells you what you want. For example `a` would have height of 32 with font size of 50. `foo` would have 42.6 and `fop` is 50. █ is 52. Indeed new line proceeding after empty line seems to have 0.5 size. I guess it is good enough value.

**Comment by kasper93** [comment]
Current version works fine. It's not readable if you don't know what is going on, but I don't wanna to bikesheed it. :)

File: player/lua/console.lua (hunk 1 above)
Comment (guidocella) [reply player/lua/console.lua]: Well yeah it changes with the exact text, it's an approximation. Japanese text is much taller and breaks the calculation, and the top line with the counter is often clipped, though with this commit it is accurate for the bottom lines at least.

**Review by kasper93 — APPROVED** [review]

**Review by kasper93 — APPROVED** [review]

**Pull request merged**